cmd/runtimetest/main.go

```diff
@@ -320,6 +322,95 @@ func validateROPaths(spec *rspec.Spec) error {
 	return nil
 }
 
+func getIDMappings(idType string) ([]rspec.IDMapping, error) {
+	var mapFile string
+	var idMaps []rspec.IDMapping
+	switch idType {
+	case "uid_map":
+		mapFile = "/proc/1/uid_map"
+	case "gid_map":
+		mapFile = "/proc/1/gid_map"
+	}
+	f, err := os.Open(mapFile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return nil, err
+		}
+
+		text := strings.Replace(strings.TrimSpace(s.Text()), "          ", " ", -1)
+		idMap := strings.Split(text, " ")
+		if len(idMap) == 3 {
+			hostID, err := strconv.ParseUint(idMap[0], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			containerID, err := strconv.ParseUint(idMap[1], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			mapSize, err := strconv.ParseUint(idMap[2], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			idMaps = append(idMaps, rspec.IDMapping{HostID: uint32(hostID), ContainerID: uint32(containerID), Size: uint32(mapSize)})
+		} else {
+			return nil, fmt.Errorf("invalid format in %v", mapFile)
+		}
+	}
+
+	return idMaps, nil
+}
+
+func validateUIDMappings(spec *rspec.Spec) error {
+	logrus.Debugf("validating uidMappings")
```

Comment:
On 09/14/2016 02:32 PM, W. Trevor King wrote:
> I think we need a guard on this for when |spec.Linux.UIDMappings| is empty. When you're creating a new user namespace, you expect |uid_map| to be empty if |UIDMappings| is empty. When you're /not/ creating a new user namespace, you expect the host's |uid_map| to be unchanged if |UIDMappings| is empty and the config is invalid <https://github.com/opencontainers/runtime-spec/blame/v1.0.0-rc1/config-linux.md#L41> if |UIDMappings| is not empty.

In my opinion:
This validation mainly intends to check if idMappings are applied as expected.
If idMappings is not empty, we should check if they are all in *_map and only them.
If not, we don't need to care if a new user namespace is created, just return error.

if idMappings is empty, this is a hard case.
As, I think we don't have ability to check if host *_map is unchanged or not.